mlir/tools/rocmlir-gen/rocmlir-gen.cpp

```diff
@@ -1268,6 +1267,152 @@ getConv2dBounds(rock::ConvOpType dir,
+  auto dst = block->getArgument(1);
+
+  auto insertConversionLogic = [&](OpBuilder &opBuilder, Value loadOp) {
+    Value newLoadOp = loadOp;
+    if (srcElemType != dstElemType) {
+      // insert conversion logic
+      auto srcBitWidth = srcElemType.getIntOrFloatBitWidth();
+      auto dstBitWidth = dstElemType.getIntOrFloatBitWidth();
+      if (srcElemType.isIntOrIndex()) {
+        if (dstElemType.isIntOrIndex()) {
+          if (srcBitWidth < dstBitWidth)
+            newLoadOp =
+                opBuilder.create<arith::ExtSIOp>(loc, dstElemType, loadOp);
+          else
+            newLoadOp =
+                opBuilder.create<arith::TruncIOp>(loc, dstElemType, loadOp);
+        } else {
+          assert(dstElemType.isa<FloatType>());
+          newLoadOp =
+              opBuilder.create<arith::SIToFPOp>(loc, dstElemType, loadOp);
+        }
+      } else {
+        assert(srcElemType.isa<FloatType>());
+        if (dstElemType.isIntOrIndex()) {
+          newLoadOp =
+              opBuilder.create<arith::FPToSIOp>(loc, dstElemType, loadOp);
+        } else {
+          if (srcBitWidth < dstBitWidth)
+            newLoadOp =
+                opBuilder.create<arith::ExtFOp>(loc, dstElemType, loadOp);
+          else
+            newLoadOp =
+                opBuilder.create<arith::TruncFOp>(loc, dstElemType, loadOp);
+        }
+      }
+    }
+    return newLoadOp;
+  };
+
+  if (srcType.getRank() == 1) {
+    auto cst0Op = b.create<arith::ConstantIndexOp>(loc, 0);
+    auto cst1Op = b.create<arith::ConstantIndexOp>(loc, 1);
+    auto size = b.create<memref::DimOp>(loc, src, cst0Op);
+
+    auto loop0 = b.create<scf::ForOp>(loc, cst0Op, size, cst1Op);
+    auto bt0 = OpBuilder::atBlockTerminator(loop0.getBody());
+    auto iv0 = loop0.getInductionVar();
+
+    Value loadOp = bt0.create<memref::LoadOp>(loc, src, ValueRange{iv0});
+    loadOp = insertConversionLogic(bt0, loadOp);
+    bt0.create<memref::StoreOp>(loc, loadOp, dst, ValueRange{iv0});
+  } else {
+    Value loadOp = b.create<memref::LoadOp>(loc, src, ValueRange{});
+    loadOp = insertConversionLogic(b, loadOp);
+    b.create<memref::StoreOp>(loc, loadOp, dst, ValueRange{});
+  }
+
+  b.create<func::ReturnOp>(loc, ValueRange{});
+
+  return func;
+}
+
+static void emitMemcpy(OpBuilder &b, Value src, Value dst) {
+  auto module = b.getBlock()->getParentOp()->getParentOfType<ModuleOp>();
+  auto loc = b.getUnknownLoc();
+
+  Value srcFlat = makeNDMemRef(b, src, 1);
+  Value dstFlat = makeNDMemRef(b, dst, 1);
+  auto srcFlatType = srcFlat.getType().cast<MemRefType>();
+  auto dstFlatType = dstFlat.getType().cast<MemRefType>();
+
+  if (srcFlatType == dstFlatType) {
+    b.create<memref::CopyOp>(loc, srcFlat, dstFlat);
+  } else {
+    auto memcpyFunc = getMemcpyFuncDecl(module, srcFlatType, dstFlatType);
+    b.create<func::CallOp>(loc, memcpyFunc, ValueRange{srcFlat, dstFlat});
+  }
+}
+
+Value ensureFloatIsF32(OpBuilder &b, Location loc, Value ref, Type floatType) {
```

Comment:
If the ref is float and not F32, make an F32 buffer and copy into it.

Reply:
Instead of commenting this here, could you make this a comment on the function and mention briefly what the function does?